### DISCUSSION TITLE
Special Value Modulators

### DISCUSSION THREAD

**Discussion by thestudiohades** [RomanKubiak/ctrlr#317] [Q&A]
Hi all,
I'm trying to design my first Vst interface for a little hardware synth i have, but i'm facing an issue regarding the modulator configuration when using uiImageSlider.

My synth have many parameters with exotic min and max values (ie. -25 to +25, -24 to +12, 0 to 255, etc.) this values are displayed on the Synth screen and all needs a 0 to 127 variation to be able to display correctly that full range whatever the min and max values are  (ie. 0=0 to 127=255,   0=-24 to 127=+12,     0=-25 to 127=+25) .

The problem is that i want to use a display on the interface (uiLCDLabel) to display the modulator name and its value. That works well with the function name displayed and the value of the modulator as well but between 0 and 127 (the value sent to the synth). The question is how to get the value as shown on the synth display (ie. -24 to +12 or others) ? It is working well for an example i found with values between -63 and +63 by setting uiSliderMin to -63, uiSliderMax to 63 and modulatorvalue+64 because the total range is 128 but i'm unable to do the same with other mix max intervals.

Did i missed a point or is this impossible without using Lua code ? If so, how to do it ?

Any help is welcome...

Eric

**Comment by dnaldoog** [comment]
Does this example panel answer the question? It uses _uiFixedSlider_ to display an interpolated value to screen while sending CC 0-127.

If you need help with the formula I can post the lua code for how to do this. I worked it out on Repl.it and could either share the page or give you the code to do yourself! You wouldn't need to use that lua code in your panel (_although I do_); just use it to calculate the interpolated values, which you would then paste into the **uiFixedSlider** _Slider Contents_ field

See attached panel!

[Interpolate using uiFixedSlider_0_0.zip](https://github.com/RomanKubiak/ctrlr/files/6613223/Interpolate.using.uiFixedSlider_0_0.zip)

**Reply by thestudiohades** [reply]
Hi John, Thanks for your fast reply !

Yes it seem to do the job ;)

I will have  a look on how it works and adapt it to my other values.

By the way, what %x means... I have never seen it before, I have already found %n, %v, %h and %s. Thinking there is maybe more but didn't find any list for that with the explanation for each, do you have something about that ?

I will be happy to get the code or the link to the page. If needed, I will use the code, i'm trying to do the all interface just using Ctrlr without using Lua ( for now at least, all I have dones work with simple functions...)

Thanks and regard

Eric

**Reply by dnaldoog** [reply]
Hi Eric!

%x prints the sysex/cc message. I think that's about it.

I am at work now and thought the repli.it would have saved. Maybe I wasn't logged in, but will post when I get home. It's basically a class I wrote so you don't need to waft through the code so much.

You will create a new object say:

<code>local x=Interpolate:new(value,1,1,-25,25)</code>

The first argument is the modulator value:

1. panel:getModulatorByName("mymod"):getValue()
2. use math.floor()=1 math ceil()=0 to round down or up
3. multiply result by 1 (in this case) : sometimes 0.10 for fractional values
4. minimum value -25
5. maximum value +25

and you can see the interpolated result using x:print()

so if you put the class though a loop, you can work out your values needed for _uiFixedSlider_.

<pre>
for i=0,127 do
x=Interpolate:new(i,1,1,-25,25)
console(tostring(x:print())
end
</pre>

Something like that!

Will post the actual code later! :)

**Comment by dnaldoog** [comment]
@thestudiohades Hi Eric,

Here is a panel that should do what you need. Let me know if it works. Not fully tested.

![bivg](https://user-images.githubusercontent.com/76831683/121208775-465bf700-c8b5-11eb-97c6-327edd95b469.png)

[Button Interpolate  Value Generator_1_0_0_2021-06-08_23-02.zip](https://github.com/RomanKubiak/ctrlr/files/6617570/Button.Interpolate.Value.Generator_1_0_0_2021-06-08_23-02.zip)

**Comment by thestudiohades** [comment]
Hello John,

Thanks again, yes it works perfectly and match with the values on the synth
;)

I face now another issue that is how to use a display and make it only
display the values from a one and only modulator (by now it display any
modulators i tweak ?

Right now I'm looking for a PDF tutorial about Ctrlr, i read on a forum
that there is one made by  Goodweather... and for some solution on the
forums ;)

Cordialy,

Eric


Le mar. 8 juin 2021 à 16:58, John Goodland ***@***.***> a
écrit :

**Comment by dnaldoog** [comment]
Hi Eric!

It gets much more complex now as I see it, because you need lua, but have a look at the source code for this panel I attach and you will see how it is done (one way). The way I wrote it allows you to easily add controls or remove controls that will print to the LCDpanel by adding to a table t{} in the callback function. 

Regards,

John

This is the function that runs when the value is changed
![callback](https://user-images.githubusercontent.com/76831683/121440056-1a7a6780-c9c2-11eb-8a15-4706feb8a2c3.png)

uncheck
![uncheckthis](https://user-images.githubusercontent.com/76831683/121440060-1d755800-c9c2-11eb-9881-e7b84cfea5db.png)

----

[Interpolate using uiFixedSlider LUA_0_0_2021-06-10_07-34.zip](https://github.com/RomanKubiak/ctrlr/files/6627329/Interpolate.using.uiFixedSlider.LUA_0_0_2021-06-10_07-34.zip)

**Comment by thestudiohades** [comment]
Thank you John, I will try to figure out how it works, at the same time I
have to read the tutorial by Goodweather... a lot of work to face off !

I'm a bit surprised by this tutorial and what I can find inside, i was
expecting the basics, such as description of each kind of modulator, the
differences between them and for what to use them... but this part is still
missing for now... anyway, i will still going on, step by step, as i'll
have to wait for the complete midi implementation chart that is not yet
written (Only CC for now...) to make this interface completely functional.
;)

Best regards,

Eric


Le jeu. 10 juin 2021 à 01:03, John Goodland ***@***.***> a
écrit :

**Reply by dnaldoog** [reply]
Hi there Eric,

There is no real documentation, except for Goodweather's excellent guide, but the old [forum ](http://ctrlr.org/forums/) has a ton of information where even the most basic questions are asked (_because there never was any real documentation to speak of_).

So I recommend looking at other panels and the **DEMO** section that would have been installed with Ctrlr. I probably learnt more things there than anywhere almost.

One thing I will say though is that nearly anything you need/want to do is possible and I continue to discover new things that I never thought possible, so it is worth persevering!

**Reply by dobo365** [reply]
Dominique (goodweather) here...
Unfortunately what you are asking and what many are expecting is a lot of work (documenting each modulator what it can do...).
I'm not the creator of Ctrlr and just a user as you. We all started knowing nothing ;-)
I started with the guide to help others but now I'm only adding a piece when I'm "in the mood".
I have now divided the new 2.x guide in 3 parts and will try to issue it part by part.
The first part is a revamped of what I already did but focusing on my Pro-2 panel io Sub 37.
The second part will cover Lua programming. I have already added a lot there but it is far from complete, at least as I want for a first release.
The third part is planned to be a reference guide. I have also quite some info there but missing many definitions. This is also the least prioritized one as not sure how many will read/need it.
Then you have nice guys like John and others who are gentle enough to add info on things I didn't cover and that I think are interesting to know so I'm also adding those (layers for example) based on their description.

As mentioned by dnaldoog/John, the bible and Ctrlr documentation is spread all over the old and now new fora. Take the time to search, look at demo and other panels and with patience you will succeed. 
And John made already many small / easy to use panels including stuff that people are looking for.

**Comment by thestudiohades** [comment]
Hello John and Dominique,

I totally agree with you, and for the past two weeks, I have spent many
hours reading Ctrlr and GitHub fora, i have also found the post with the
"future" of your tutorial (the word 2.5 version with a lot of Lua
explanations, and also downloaded panels that have the function i need to
do (or close at). I already started to see things clearer and I must thank
you for the boost , taking some of your time to help me or giving me some
panels.

I will now have time to improve my skill, as all that can be done on my
interface is already done, and that I must now wait for the complete MIDI
implementation, to be written by the designer of this synth.. in a few
months...

Thanks again 😉

Eric

Le jeu. 10 juin 2021 à 18:16, goodweather ***@***.***> a
écrit :
